LOGGING.md

```diff
@@ -0,0 +1,58 @@
+# Logging
+
+This file aims to provide a sort of standardization for our logging. Uniformity allows easier
+processing after-the-fact (e.g. via LogQL in Grafana), so having formally defined norms can help to
+reduce the cognitive overhead of writing new code or making changes to existing code.
+
+#### Components checklist
+
+The following components have been updated to follow this document:
+
+- [x] autoscaler-agent
+- [x] autoscale-scheduler (scheduler plugin)
+- [x] vm-informant
+- [ ] neonvm-controlller
+- [ ] neonvm-runner
+
+## Common keys
+
+- `pod: NamespacedName` — the Pod related to some operation (or long-running handler)
+- `virtualmachine: NamespacedName` — the VirtualMachine related to some operation (or long-running
+  handler). **Note:** Wherever possible, `pod` should be set alongside this field, or `null` if the
+  VM's `.status.podName` field is not set.
+- `node: string` — name of the k8s node that a particular pod or VM is on. Only applicable for some
+    components.
+- `error: string` — for error-level and above, the error that occurred.
+    - This is typically handled by `zap.Error(err)`, so no extra naming is necessary.
+
+## Logger name structuring
+
+Zap loggers have names. For example, a `zap.Logger` with the name `autoscaler-agent.watch` will
+output lines like:
+
+```
+{"level":"info","logger":"autoscaler-agent.watch","msg":"hello from the logger!"}
+```
+
+Names are hierarchical — creating a new logger with [`logger.Named(name)`] appends `.{name}` to the
+`logger` field for lines generated by the new logger.
+
+[`logger.Named(name)`]: https://pkg.go.dev/go.uber.org/zap#Logger.Named
+
+_In general_, all tight groupings of long-lived goroutines (sometimes just one) are given a leaf
+node in the hierarchy. For example: the autoscaler-agent's `Runner`s spawn a handful of groroutines
+to handle various tasks. The goroutine responsible for updating which scheduler to communciate with
+has limited communication and a lifecycle that is notably independent from other parts of the
+`Runner`. But the part that fetches metrics is (relatively) closely tied to the part that updates
+the VM's resources, so they share a logger name (`agent.runner.main`).
```

Comment:
To be honest, hierarchical logging functionality is perhaps one of my least favorite and most bothersome logging features (and I think it's kind of retro and imagines a different sort of execution model and way of interaction with logging from the one we actually have/interact with these days.) 

In short it makes a lot of sense as a way of filtering logs, particularly when you might want to control log output at runtime (e.g. moving from info to debug transiently) or to be able to set some hierarchies with different levels. But we don't do this, and also often use much better tools (like grep! or jq! or splunk!) to search and filter logs. 

What you're left with is this "magic", potentially easy to parse, string that is supposed to tell you about the component that you end up filtering things by, and then if a component moves around or gets called from a different place, you can end up with a different hierarchy and therefore a different string. In Cloud (and this is an incredibly common  mode, we create new loggers in the hierarchy using `logger.FromLogger(logger, "workerOp")` or similar... and as things move around or get called from a different place the messages end up changing. 

We have this in cloud and it's _it's confusing_. Sometimes operations (e.g. applying a config) are triggered by background jobs, in which case the log hierarchy says "background job" at the top, but sometimes we call the same code from a web request, which means that the loggers get `publicapi_v2` as a prefix. 

We can say "don't do this" or "be very careful about creating new loggers," but adding friction to logging, means that people will be less likely to use loggers or will do something slapdash when firefighting, and the point of using better logging tools is to avoid this kind of friction and hesitation. 

-----

I've always thought that the combination of structured logging and hierarchical logging doesn't make a lot of sense. Just put the component name or system name in as a field, throw all logs into one stream and filter it out later as you need it. Structured logging gives you the ability, in practice, to not need to think about this at all. 

-----

One thing that hierarchical logging kind of gives you that's hard to recreate directly with structured logging is the _sense_ of a call stack for a logging component. This can be quite useful, but I'm inclined to acctually just have helpers that annotate specific log messages (or all log messages depending on volume or form.) with stack traces.

Reply:
recording a proper response here for posterity (even though we talked through most of this already):

> To be honest, hierarchical logging functionality is perhaps one of my least favorite and most bothersome logging features

:disappointed:

> I think [hierarchical logging is] kind of retro and imagines a different sort of execution model and way of interaction with logging from the one we actually have/interact with these days.

I can see how the execution model piece makes sense for something like the control plane. For this repo, and _particularly_ for the design of the autoscaler-agent, we _do_ actually have many independent goroutines doing stuff — without a central jobs distributor managing them.

> In short it makes a lot of sense as a way of filtering logs, particularly when you might want to control log output at runtime (e.g. moving from info to debug transiently) or to be able to set some hierarchies with different levels. But we don't do this, and also often use much better tools (like grep! or jq! or splunk!) to search and filter logs.

A few things here:

1. I wanted it for the filtering! Filtering is good. The existing autoscaling logs uses filename-based filtering to achieve what I'd hope would be accomplished by this, and even though it's very hacky, it's _incredibly_ useful.
2. Controlling log output at runtime is... fine, I guess? I have no desire to do that here.
3. Relating to tools, IME we do not currently use much better tools (see for example the [autoscaling logs dashboard](https://neonprod.grafana.net/d/cc4f137c-f2fd-4e18-a00a-c92234e3b33a/neon-logs-autoscaling)); and even if we did, the ability to filter by the internal subsystem is always going to be difficult if we don't _explicitly_ make that information available in the logs.

> What you're left with is this "magic", potentially easy to parse, string that is supposed to tell you about the component that you end up filtering things by, and then if a component moves around or gets called from a different place, you can end up with a different hierarchy and therefore a different string. In Cloud (and this is an incredibly common mode, we create new loggers in the hierarchy using `logger.FromLogger(logger, "workerOp")` or similar... and as things move around or get called from a different place the messages end up changing.
>
> We have this in cloud and it's it's confusing. Sometimes operations (e.g. applying a config) are triggered by background jobs, in which case the log hierarchy says "background job" at the top, but sometimes we call the same code from a web request, which means that the loggers get `publicapi_v2` as a prefix.

Regarding the top-level loggers in cloud, from what you've said it does seem that the `logger` field is unhelpful under the current architecture. See also: the stuff I said above about the architecture in this repo being different.

In my view, this is not something that should be a "magic" string in any sense; there should be a fixed number of `logger` values, so that you can filter over which subsystems you'd like to view the output for (or gather metrics by "which subsystem is experiencing errors", etc.). This is why in my examples it's things like `autoscaler-agent.runner.main`.

> We can say "don't do this" or "be very careful about creating new loggers," but adding friction to logging, means that people will be less likely to use loggers or will do something slapdash when firefighting, and the point of using better logging tools is to avoid this kind of friction and hesitation.

I don't think there's any (or: at least not many) shortcuts to get to a place where the structured logging is easy to manipulate algorithmically — it requires common keys, and various other disciplined behaviors that are _easier_ with a better-designed architecture, but still aren't guaranteed! Some of this can be fixed by investing into an internal logging shim/API that provides a lot of these "common keys" as methods on the logger as a form of standardization. Ultimately I'm skeptical of the idea that there's low-effort ways to get high-quality logs.

> I've always thought that the combination of structured logging and hierarchical logging doesn't make a lot of sense. Just put the component name or system name in as a field, throw all logs into one stream and filter it out later as you need it. Structured logging gives you the ability, in practice, to not need to think about this at all.
>
> One thing that hierarchical logging kind of gives you that's hard to recreate directly with structured logging is the sense of a call stack for a logging component. This can be quite useful, but I'm inclined to acctually just have helpers that annotate specific log messages (or all log messages depending on volume or form.) with stack traces.

A final couple notes:

1. I think under the model of "logger names are a proxy for a callstack" (which you elude to here), the hierarchical logging thing _does_ fall apart. I'm not intending for the represented hierarchy to represent a callstack (nor does LOGGING.md say that). My belief is that the set of guidelines I've proposed should work reasonably for this project as it currently is, and avoid some of the pitfalls you've described.
2. The immediate question in response to "just put the component name as a field" is: "given the existing hierarchy, why aren't you just using the `.Named()` method to do this for you?". There's value in using the standardized thing, even if the API is... lacking, and doesn't grant nearly as much freedom to change the `logger` name as we'd like.